src/main/resources/org/jenkinsci/plugins/docker/workflow/Docker.groovy

```diff
@@ -159,8 +159,15 @@ class Docker implements Serializable {
         public void tag(String tagName = parsedId.tag, boolean force = true) {
             docker.node {
                 def taggedImageName = toQualifiedImageName(parsedId.userAndRepo + ':' + tagName)
-                // TODO as of 1.10.0 --force is deprecated; for 1.12+ do not try it even once
-                docker.script.sh "docker tag --force=${force} ${id} ${taggedImageName} || docker tag ${id} ${taggedImageName}"
+                //get docker version
+                def dockerVersion = docker.script.sh(script: "docker version --format '{{.Client.Version}}'", returnStdout: true).trim().split('.')
+                def dockerMainVersion = dockerVersion[0].isInteger() ? dockerVersion[0].toInteger() : null
+                def dockerMinorVersion = dockerVersion[1].isInteger() ? dockerVersion[1].toInteger() : null
+                if (dockerMainVersion >= 1 && dockerMinorVersion >= 12) {
+                    docker.script.sh "docker tag ${id} ${taggedImageName}"
+                } else {
+                    docker.script.sh "docker tag --force=${force} ${id} ${taggedImageName} || docker tag ${id} ${taggedImageName}"
```

Comment:
The ```||``` may not be adequate here, because when running the tag command on docker 1.12, it will fail and return a non-zero error code. If the shell process is running in the `set -e` mode, that error will be immediately propagated.

My results seem to suggest that the error 125 that docker returns in the case where an unsupported ```--force=true``` is present is causing a push operation to fail.

```
mtolan-retina:docker-workflow-plugin mtolan$ docker tag --force=true 7b0b80cd7743 myimage:foo
unknown flag: --force
See 'docker tag --help'.
mtolan-retina:docker-workflow-plugin mtolan$ echo $?
125
```

```
Entering stage Push Image to ECR
Proceeding
[Pipeline] withEnv
[Pipeline] {
[Pipeline] withDockerRegistry
Wrote authentication to /var/lib/jenkins/.dockercfg
[Pipeline] {
[Pipeline] sh
[prlic] Running shell script
+ docker tag --force=true playready-license 410400486419.dkr.ecr.us-east-1.amazonaws.com/playready-license:b6860ccf6a4c
unknown flag: --force
See 'docker tag --help'.
[Pipeline] }
[Pipeline] // withDockerRegistry
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 125
Finished: FAILURE
```

In any event, relying on the ```||``` to ensure that things work successfully is probably not the best approach if we expect it to fail. At _best_ it will result in additional misleading log lines that likely result in confusion on other tickets.